### PR TITLE
TFA-FIX:CEPH-83595932-To verify crashes while executing drain and mgr failover commands

### PR DESCRIPTION
**Pull request by SrinivasaBharath**
# Description

In test case execution failed and the failed log is -
 http://magna002.ceph.redhat.com/cephci-jenkins/results/openstack/RH/8.0/rhel-9/Regression/19.2.0-52/rados/45/tier-2_rados_test-drain-customer-issue

Jira tasks to track the issue are -
1. https://issues.redhat.com/browse/RHCEPHQE-16628
2. https://issues.redhat.com/browse/RHCEPHQE-16835 

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by SrinivasaBharath** [comment]
Multiple times I tested the scenario and the following are the pass logs-

http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-6843V9/ 
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-ZFAZ0D 
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-GBP28Y 
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-I5DB23 
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-YW1ITT
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-BK83CY
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-083IN3
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-XGXKVJ
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1FM1QR

File: ceph/rados/serviceability_workflows.py [hunk 1]
```diff
@@ -225,22 +225,71 @@ def remove_custom_host(self, host_node_name: str):
...
             None | raises exception in case of failure
         """
+        status_cmd = ""
         try:
+
+            def wait_osd_operation_status(status_cmd):
+                status_flag = False
+                end_time = datetime.datetime.now() + datetime.timedelta(seconds=600)
+                while end_time > datetime.datetime.now():
+                    out, err = self.cephadm.shell([status_cmd])
```
Comment (pdhiran) [inline ceph/rados/serviceability_workflows.py:235]: Any reason for executing commands with `self.cephadm.shell()` and then performing operations on data?

File: ceph/rados/serviceability_workflows.py [hunk 2]
```diff
@@ -225,22 +225,71 @@ def remove_custom_host(self, host_node_name: str):
...
+                    try:
+                        drain_ops = json.loads(out)
+                        for entry in drain_ops:
+                            log.debug(
+                                f"OSD remove operation is in progress {osd_id}\nOperations: {entry}"
+                            )
+                    except json.JSONDecodeError:
+                        log.info(f"The OSD removal is completed on OSD : {osd_id}")
```
Comment (pdhiran) [inline ceph/rados/serviceability_workflows.py:243]: This could either mean that OSD removal is complete, or No OSD removal was started in the 1st place.

File: ceph/rados/serviceability_workflows.py (hunk 1 above)
Comment (pdhiran) [reply ceph/rados/serviceability_workflows.py:235]: Please add reason for `self.cephadm.shell()` and then performing operations on data

File: tests/rados/test_node_drain_customer_bug.py [hunk 3]
```diff
@@ -152,6 +158,12 @@ def run(ceph_cluster, **kw):
                 "The traceback messages are noticed in logs.The error snippets are noticed in the MGR logs"
             )
             return 1
+        cmd_unset_unmanaged = (
```
Comment (pdhiran) [inline tests/rados/test_node_drain_customer_bug.py:161]: Same as above. let's use the new method that will be created.

**Review by pdhiran — COMMENTED** [review]
Overall looks good.

New method to set services to managed and unmanaged to true/ false should be created and used here.

File: tests/rados/test_node_drain_customer_bug.py (hunk 3 above)
Comment (SrinivasaBharath) [reply tests/rados/test_node_drain_customer_bug.py]: Used new method.

File: tests/rados/test_node_drain_customer_bug.py [hunk 4]
```diff
@@ -135,6 +135,12 @@ def run(ceph_cluster, **kw):
         log.info(
             f"The OSDs in the drain node before starting the test - {osd_count_before_test} "
         )
+        cmd_set_unmanaged = (
```
[the remark this answers is not in the document]
Comment (SrinivasaBharath) [reply tests/rados/test_node_drain_customer_bug.py]: New method is created and included.

File: ceph/rados/serviceability_workflows.py (hunk 1 above)
Comment (SrinivasaBharath) [reply ceph/rados/serviceability_workflows.py:243]: The information is added.

**Comment by SrinivasaBharath** [comment]
Preempt scrub fix pass log- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-HWPGNS

File: ceph/rados/utils.py [hunk 5]
```diff
@@ -45,7 +45,6 @@ def set_osd_devices_unmanaged(ceph_cluster, osd_id, unmanaged):
             break
 
     if not service_name:
-        log.error(f"No orch service found for osd: {osd_id}")
```
Comment (harshkumarRH) [inline ceph/rados/utils.py:48]: @SrinivasaBharath please explain why we are making this change?

File: ceph/rados/core_workflows.py [hunk 6]
```diff
@@ -4615,3 +4615,60 @@ def get_rados_df(self, pool_name: str = None):
...
+        if unmanaged:
+            log.debug(
+                f"Setting the {service_type} service as unmanaged by cephadm. current status : {out}"
+            )
+            out["unmanaged"] = "true"
+        else:
+            log.debug(
+                f"Setting the {service_type} service as unmanaged by cephadm. current status : {out}"
+            )
+            out["unmanaged"] = "false"
```
Comment (harshkumarRH) [inline ceph/rados/core_workflows.py:4640]: Here instead of explicitly setting it to False, we should remove the "unmanaged" key from the dictionary if it exists

File: tests/rados/test_node_drain_customer_bug.py [hunk 7]
```diff
@@ -42,19 +42,20 @@ def run(ceph_cluster, **kw):
     service_obj = ServiceabilityMethods(cluster=ceph_cluster, **config)
     ceph_nodes = kw.get("ceph_nodes")
     config = kw["config"]
-
+    # cmd_unset_unmanaged = ""
     replicated_config = config.get("replicated_pool")
     pool_name = replicated_config["pool_name"]
-    active_osd_list = rados_obj.get_osd_list(status="up")
+    active_osd_list = rados_obj.get_active_osd_list()
```
Comment (harshkumarRH) [inline tests/rados/test_node_drain_customer_bug.py:48]: Please revert the change in this line to "active_osd_list = rados_obj.get_osd_list(status="up")"
The get_active_osd_list() method no longer exists

File: tests/rados/test_node_drain_customer_bug.py [hunk 8]
```diff
@@ -133,8 +134,10 @@ def run(ceph_cluster, **kw):
     try:
         osd_count_before_test = get_node_osd_list(rados_obj, ceph_nodes, drain_host)
         log.info(
-            f"The OSDs in the drain node before starting the test - {osd_count_before_test} "
+            f"st The OSDs in the drain node before starting the te- {osd_count_before_test} "
```
Comment (harshkumarRH) [inline tests/rados/test_node_drain_customer_bug.py:137]: Please fix typo

File: tests/rados/test_node_drain_customer_bug.py [hunk 9]
```diff
@@ -194,7 +200,7 @@ def run(ceph_cluster, **kw):
             return 1
 
         if bug_exists:
-            active_osd_list = rados_obj.get_osd_list(status="up")
+            active_osd_list = rados_obj.get_active_osd_list()
```
Comment (harshkumarRH) [inline tests/rados/test_node_drain_customer_bug.py:203]: Please revert to using get_osd_list(status="up")

File: tests/rados/test_rados_preempt_scrub.py [hunk 10]
```diff
@@ -85,7 +85,7 @@ def run(ceph_cluster, **kw):
 
             log_lines = [
                 "head preempted",
-                "WaitReplicas::react(const GotReplicas&) PREEMPTED",
+                "WaitReplicas::react(const GotReplicas&) PREEMPTED!",
```
Comment (harshkumarRH) [inline tests/rados/test_rados_preempt_scrub.py:88]: @SrinivasaBharath I request more clarity here.
The comparison happening in `def verify_preempt_log` is not == rather `in`
So previously if a subset of the sentence was not being found in log lines, how is adding an additional character '!' going to make any difference?

File: tests/rados/test_node_drain_customer_bug.py (hunk 8 above)
Comment (SrinivasaBharath) [reply tests/rados/test_node_drain_customer_bug.py:141]: Fixed the typo

File: tests/rados/test_node_drain_customer_bug.py (hunk 7 above)
Comment (SrinivasaBharath) [reply tests/rados/test_node_drain_customer_bug.py]: get_active_osd_list method is the local method  combines the in and up osd list

File: tests/rados/test_node_drain_customer_bug.py (hunk 9 above)
Comment (SrinivasaBharath) [reply tests/rados/test_node_drain_customer_bug.py]: Done.

File: ceph/rados/core_workflows.py (hunk 6 above)
Comment (pdhiran) [reply ceph/rados/core_workflows.py:4640]: This should be fine. Would not make difference with workflow.

File: tests/rados/test_node_drain_customer_bug.py [hunk 11]
```diff
@@ -40,21 +42,23 @@ def run(ceph_cluster, **kw):
...
     service_obj = ServiceabilityMethods(cluster=ceph_cluster, **config)
+    mon_obj = MonConfigMethods(rados_obj=rados_obj)
     ceph_nodes = kw.get("ceph_nodes")
     config = kw["config"]
-
+    # cmd_unset_unmanaged = ""
     replicated_config = config.get("replicated_pool")
     pool_name = replicated_config["pool_name"]
-    active_osd_list = rados_obj.get_osd_list(status="up")
+    active_osd_list = get_active_osd_list(rados_obj)
```
Comment (pdhiran) [inline tests/rados/test_node_drain_customer_bug.py:51]: reuse existing method

File: ceph/rados/utils.py (hunk 5 above)
Comment (SrinivasaBharath) [reply ceph/rados/utils.py:48]: It is a mistake from my side. I reverted the changes.

File: tests/rados/test_node_drain_customer_bug.py (hunk 11 above)
Comment (SrinivasaBharath) [reply tests/rados/test_node_drain_customer_bug.py]: Used the existing method.

File: ceph/rados/core_workflows.py [hunk 12]
```diff
@@ -4615,3 +4615,71 @@ def get_rados_df(self, pool_name: str = None):
         out = self.run_ceph_command(cmd=_cmd, client_exec=True)
 
         return out["pools"][0] if pool_name else out
+
+    def get_service_list(self, service_type=None):
```
Comment (harshkumarRH) [inline ceph/rados/core_workflows.py:4619]: Please check existing method `def list_orch_services(self, service_type=None) -> list`

Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py]: Updated the code.

File: tests/rados/test_rados_preempt_scrub.py (hunk 10 above)
Comment (SrinivasaBharath) [reply tests/rados/test_rados_preempt_scrub.py]: I reverted the changes. The fix will be provided in a separate PR.

**Comment by SrinivasaBharath** [comment]
Pass log with Reef- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-N6PKVO/

File: ceph/rados/core_workflows.py [hunk 13]
```diff
@@ -4615,3 +4616,56 @@ def get_rados_df(self, pool_name: str = None):
...
+        """
+        Method to set the service to either managed or unmanaged
+        Args:
+            unmanaged: True or false, for the service management
+            service_type : service types are- mon,mgr,osd,rgw, mds
+        returns:
+            Pass -> True, Fail -> false
+        """
+        cmd_export = f"ceph orch ls {service_type} --export"
+        out = self.run_ceph_command(cmd=cmd_export, client_exec=True)[0]
```
Comment (pdhiran) [inline ceph/rados/core_workflows.py:4630]: We should be retrieving all the items in list.
Eg : 

```# ceph orch ls osd --export -f json-pretty

[
  {
    "placement": {
      "label": "osd"
    },
    "service_id": "osds_colocated",
    "service_name": "osd.osds_colocated",
    "service_type": "osd",
    "spec": {
      "data_devices": {
        "paths": [
          "/dev/ssd-vg/ssd-lv1"
        ]
      },
      "filter_logic": "AND",
      "objectstore": "bluestore"
    }
  },
  {
    "placement": {
      "label": "osd"
    },
    "service_id": "osds_noncolocated",
    "service_name": "osd.osds_noncolocated",
    "service_type": "osd",
    "spec": {
      "data_devices": {
        "paths": [
          "/dev/hdd-vg/hdd-lv1",
          "/dev/hdd-vg/hdd-lv2",
          "/dev/hdd-vg/hdd-lv3"
        ]
      },
      "db_devices": {
        "paths": [
          "/dev/nvme-vg/nvme-lv1",
          "/dev/nvme-vg/nvme-lv2",
          "/dev/nvme-vg/nvme-lv3"
        ]
      },
      "filter_logic": "AND",
      "objectstore": "bluestore"
    },
    "unmanaged": true
  }
] ```

File: ceph/rados/core_workflows.py [hunk 14]
```diff
@@ -4615,3 +4616,56 @@ def get_rados_df(self, pool_name: str = None):
...
+        # Adding the spec rules into the file
+        cmd = f"echo '{json_out}' > {file_name}"
+        self.client.exec_command(cmd=cmd, sudo=True)
+
+        log.debug(f"Contents of {service_type} spec file : {out}")
+        apply_cmd = f"ceph orch apply -i {file_name}"
+        log.info(f"Applying the spec file via cmd : {apply_cmd}")
+        self.client.exec_command(cmd=apply_cmd, sudo=True)
+        time.sleep(10)
+        out = self.list_orch_services(service_type="osd", export=True)[0]
```
Comment (pdhiran) [inline ceph/rados/core_workflows.py:4657]: Same here. iterate through all the elements for the services.

Comment (pdhiran) [inline ceph/rados/core_workflows.py:4657]: Service type hardcoded as OSD. make it generalized. `{service_type}`

File: ceph/rados/serviceability_workflows.py (hunk 2 above)
Comment (s-vipin) [reply ceph/rados/serviceability_workflows.py:251]: We could use `ceph pg ls-by-osd OSD_ID` to check if PGs are drained or not.

File: ceph/rados/core_workflows.py (hunk 14 above)
Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py:4657]: The same as the above-
 /home/skanta/cephci_tfa/cephci/ceph/rados/core_workflows.py(4661)set_service_managed_type()
-> out = self.list_orch_services(service_type=service_type, export=True)[0]
(Pdb) n
2024-12-02 05:31:39,811 (cephci.test_delete) [INFO] - cephci.ceph.ceph.py:1570 - Execute ceph orch ls osd --export -f json on 10.0.64.178
2024-12-02 05:31:40,825 (cephci.test_delete) [INFO] - cephci.ceph.ceph.py:1600 - Execution of ceph orch ls osd --export -f json on 10.0.64.178 took 1.005422 seconds.
> /home/skanta/cephci_tfa/cephci/ceph/rados/core_workflows.py(4662)set_service_managed_type()
-> out_test = self.list_orch_services(service_type=service_type, export=True)
(Pdb) n
2024-12-02 05:31:43,029 (cephci.test_delete) [INFO] - cephci.ceph.ceph.py:1570 - Execute ceph orch ls osd --export -f json on 10.0.64.178
2024-12-02 05:31:44,039 (cephci.test_delete) [INFO] - cephci.ceph.ceph.py:1600 - Execution of ceph orch ls osd --export -f json on 10.0.64.178 took 1.005566 seconds.
> /home/skanta/cephci_tfa/cephci/ceph/rados/core_workflows.py(4663)set_service_managed_type()
-> status = out.get("unmanaged", False)
(Pdb) p out
{'placement': {'host_pattern': '*'}, 'service_id': 'all-available-devices', 'service_name': 'osd.all-available-devices', 'service_type': 'osd', 'spec': {'data_devices': {'all': True}, 'filter_logic': 'AND', 'objectstore': 'bluestore'}, 'unmanaged': 'false'}
(Pdb) p out_test
[{'placement': {'host_pattern': '*'}, 'service_id': 'all-available-devices', 'service_name': 'osd.all-available-devices', 'service_type': 'osd', 'spec': {'data_devices': {'all': True}, 'filter_logic': 'AND', 'objectstore': 'bluestore'}, 'unmanaged': 'false'}]
(Pdb) p type(out)
<class 'dict'>
(Pdb) p type(out_test)
<class 'list'>
(Pdb)

Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py:4657]: Changed the service_type

**Comment by SrinivasaBharath** [comment]
Pass log- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-RIOLY9

File: ceph/rados/serviceability_workflows.py (hunk 2 above)
Comment (SrinivasaBharath) [reply ceph/rados/serviceability_workflows.py:251]: Implemented the logic by using the ceph orch ps hostname command

**Comment by SrinivasaBharath** [comment]
Pass Log- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IF8FR9

File: ceph/rados/core_workflows.py (hunk 13 above)
Comment (pdhiran) [reply ceph/rados/core_workflows.py:4638]: each service needs to be set to unmanaged. eg : 

```out = json.loads(osd_orch_op)

for outr1 in out:
    outr1.setdefault("unmanaged", True)
    print(outr1)```

File: ceph/rados/serviceability_workflows.py [hunk 15]
```diff
@@ -253,36 +317,15 @@ def remove_custom_host(self, host_node_name: str):
...
-                    raise Exception(error)
-                log.debug("Sleeping for 10 seconds and checking again....")
-                time.sleep(10)
-
-            if not flag:
+            if wait_osd_operation_status(status_cmd):
+                log.info(
+                    f"Completed drain operation on the host. {rm_host.hostname}\n Removing host from the cluster"
+                )
+            else:
```
Comment (pdhiran) [inline ceph/rados/serviceability_workflows.py:324]: just before removal, check if any services exist. line no 335.

File: ceph/rados/utils.py [hunk 16]
```diff
@@ -195,17 +195,23 @@ def set_osd_in(
...
         ceph_cluster: ceph cluster
         osd_id: osd id
         zap: flag to control zapping of device
+        force: flag to remove the OSD forcefully
     """
     config = {"command": "rm", "service": "osd", "pos_args": [osd_id]}
+    cmd_args = {}
     if zap:
-        config["base_cmd_args"] = {"zap": True}
+        cmd_args["zap"] = True
```
Comment (pdhiran) [inline ceph/rados/utils.py:210]: let's give force as true when zap is true as well.

File: ceph/rados/serviceability_workflows.py (hunk 15 above)
Comment (SrinivasaBharath) [reply ceph/rados/serviceability_workflows.py:324]: Included the check.

File: ceph/rados/core_workflows.py (hunk 14 above)
Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py]: Modified the code as pre the comments.

File: ceph/rados/core_workflows.py (hunk 13 above)
Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py:4638]: Included the code as per comments.

File: ceph/rados/utils.py (hunk 16 above)
Comment (SrinivasaBharath) [reply ceph/rados/utils.py:210]: modified the code.

File: ceph/rados/core_workflows.py [hunk 17]
```diff
@@ -4615,3 +4623,57 @@ def get_rados_df(self, pool_name: str = None):
...
+            json_out = json.dumps(osd_service)
+            # Adding the spec rules into the file
+            cmd = f"echo '{json_out}' > {file_name}"
+            self.client.exec_command(cmd=cmd, sudo=True)
+            log.debug(f"Contents of {service_type} spec file : {out}")
+            apply_cmd = f"ceph orch apply -i {file_name}"
+            log.info(f"Applying the spec file via cmd : {apply_cmd}")
+            self.client.exec_command(cmd=apply_cmd, sudo=True)
+            time.sleep(10)
+            out = self.list_orch_services(service_type=service_type, export=True)
```
Comment (pdhiran) [inline ceph/rados/core_workflows.py:4665]: this method call should be outside the for loop

Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py:4665]: Moved outside the loop

**Review by pdhiran — COMMENTED** [review]
looks good

**Review by pdhiran — APPROVED** [review]
looks good
